Comments

gingin77 opened the pull request:
**What**
- I split the Submission class into 2 classes - Comment and Post; they are both using the same submissions data table
- I used Rspec to test models and set up validations for Comment and Post classes
- The seed file is very long - Should I be using FactoryBot here? I'm not sure the best way to assess what should be raked if the db is not clear
- the `posts#show` view has a lot of html to handle nested comments - there's probably a better way, but I couldn't figure it out this afternoon

gingin77 committed:
WIP

gingin77 committed:
WIP - Rpec and data-cleaner gems have been added

Factories updated for new post
A handful of passing specs have been written for the Post model

Schema now has tables for posts and comments that need be be rolled back
The Submission model will eventually be deleted

gingin77 committed:
Write tests and validations on Comments model - in tests: focus was on direct_comments, rather than nested comments. Update migrations and schema.

gingin77 committed:
Write more specs and validations for Post model - all existing specs pass

gingin77 committed:
Delete Submission model

gingin77 committed:
Update seeds file to use Post and Comment in place of Submission

gingin77 committed:
Work up new comments and posts controllers - go through steps to create new posts and comments; things are working!

gingin77 committed:
Set up association between Comments and Posts in respective models and add related helper_methods in controllers.
Define 'reply_comments' rather than 'nested_comments'.
Define nested comments#new route and use 'reply' as alias
Update factories to look for reply rather than nested
All existing tests are currently passing.

gingin77 committed:
Add html/erb to views post#show and comment#new views where replies display. Add styling to indent as temporary fix. A more sophisticated indentation strategy will be needed for extensive nesting.

gingin77 committed:
Work with nested comments - modeling and view presentation

Create more seed data examples
Rename index column comment_id, instead of direct_comment_id - run migration, so schema is up to date

gingin77 committed:
The view for a new reply comment only shows the comment being replied to and no others in chain

gingin77 committed:
Delete tests relying on Minitest - I'm switching to Rspec

gingin77 committed:
Fix bug in Factory for comments that was making tests fail

gingin77 committed:
Fix link_to command so link is active, not just displayed

gingin77 committed:
Create new seeds dir and add more data from a Facebook group that I'm an admin for - after changing all names to protect peoples' privacy.

zporter commented on Gemfile:
```diff
@@ -5,44 +5,33 @@ git_source(:github) do |repo_name|
   gem 'spring-watcher-listen', '~> 2.0.0'
 end
 
-# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
```
I personally prefer these comments in a Gemfile. Since these gems don't indicate what they are in their name (and could be yanked at any time without warning), it's helpful to keep this information with the project. I'm sad to see them go.

gingin77 replied:
Hmm... that makes sense. I don't want you to be sad - I can bring them back!

zporter commented on app/assets/stylesheets/application.scss:
```diff
@@ -111,3 +111,7 @@ a {
 }
+
+.replies-posted {
+ margin-left: 50px;
```
indentation seems off here

zporter commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def direct_comments_on_post
+    @comments ||= Comment.where(post_id: parent_post.id, comment_id: nil)
+  end
```
this method doesn't appear to be used

zporter commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,41 @@
+    else
+      flash[:inline] = @comment.errors.messages[:text].join(", ")
+      redirect_to post_path(@comment.post_id)
+    end
```
since these redirect to the same place, I'd move it outside of the conditional.

zporter commented on app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,45 @@
+    @post = Post.new(post_params.merge(submitter: current_user))
+    if @post.save
+      flash[:alert] = "Your post was saved"
+      redirect_to post_path(@post.id)
```
the `.id` is unnecessary

zporter replied:
FWIW, you can combine the above 2 lines like so:

```ruby
redirect_to @post, alert: 'Your post was saved'
```

zporter commented on app/models/comment.rb:
```diff
@@ -0,0 +1,14 @@
+  belongs_to :submitter, class_name: :User, foreign_key: :user_id
+  belongs_to :post
+
+  # scope :comments, -> { where(post_id: !nil) }
```
remove this?

gingin77 replied:
At first I wrote it in, just in anticipation that I'd probably use it, but when I was getting the PR ready, I couldn't remember if I'd tested out how it was working, so left it as a comment.

zporter commented on app/models/post.rb:
```diff
@@ -0,0 +1,38 @@
+  belongs_to :submitter, class_name: :User, foreign_key: :user_id
+  has_many :comments
+  has_many :direct_comments, -> { where(comment_id: nil) }, class_name: "Comment"
+  has_many :reply_comments, through: :direct_comments
```
can you explain the separation between direct and reply comments?

gingin77 replied:
Direct comments have a post_id, but not a comment_id; Reply_comments have both a post_id (parent post) and a comment_id(parent_comment).

At some point after pushing this up, I realized the hmat relationship only reaches one level of replies. A reply to a another reply is not captured by the statement above.

gingin77 committed:
Fix identation

gingin77 committed:
Refactor post#show view, replace submissions table with 2 tables - posts and comments

All existing specs are currently passing

zporter commented on app/controllers/application_controller.rb:
```diff
@@ -9,4 +9,12 @@ def configure_permitted_parameters
+      flash[:alert] = "You must be logged in to submit a post or comment"
+      redirect_to new_user_session_path
+    end
+  end
```
Might be worth tucking this in a `Submitable` module shared between the Posts and Comments controllers.

gingin77 replied:
Hmm - making this issue here: https://github.com/gingin77/hackernews_clone/issues/17

zporter commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,39 @@
+
+  def create
+    @comment = Comment.new(comment_params.merge(submitter: current_user))
+    if !@comment.save
```
slight preference for using `unless` over `it !`

zporter commented on app/models/post.rb:
```diff
@@ -0,0 +1,34 @@
+  belongs_to :submitter, class_name: :User, foreign_key: :user_id
+
+  has_many :comments
+  has_many :direct_comments, -> { where(comment_id: nil) }, class_name: "Comment"
```
Did you consider using [polymorphic associations](http://guides.rubyonrails.org/association_basics.html#polymorphic-associations) here?

gingin77 replied:
See thread below regarding new comment form partial [https://github.com/gingin77/hackernews_clone/pull/15#pullrequestreview-95941585](url) 

https://github.com/gingin77/hackernews_clone/issues/18

zporter commented on app/models/comment.rb:
```diff
@@ -0,0 +1,9 @@
+  belongs_to :direct_comment, class_name: "Comment", optional: true
+
+  validates :text, presence: { message: 'The comment you submitted was blank' }
+end
```
I see some common code between this and the `Post` class. Worth a `Submitable` module to share this code?

gingin77 replied:
Hmm - making this issue here: https://github.com/gingin77/hackernews_clone/issues/17

zporter commented on app/views/comments/new.html.erb:
```diff
@@ -0,0 +1,13 @@
+    <% end %>
+  <% end %>
+  <%= f.submit "Submit Reply" %>
+<% end %>
```
is this the same form that shows up on `posts#show`? If so, can you think of a way to share it?

gingin77 replied:
I've just made a list of aspects that make each form distinct from one another. 

In form_for options hash:
  new: `url: post_comments_path(parent_post.id)`
  show: `url: post_comments_path(post.id)`

Within form body:
  new: `<%= f.hidden_field :comment_id, :value => parent_comment.id %>`
  show: ^^ Does not have the line above

  new: ` <%= f.hidden_field :post_id, :value => parent_post.id %>`
  show: ` <%= f.hidden_field :post_id, :value => post.id %>`

Given that all of these are related to naming a parent, it seems like this could be a case where making Comments polymorphic as commentable could offer an advantage. From the Rails guide, "If you have an instance of the Picture model, you can get to its parent via @picture.imageable." 
[http://guides.rubyonrails.org/association_basics.html#polymorphic-associations](url)

gingin77 replied:
https://github.com/gingin77/hackernews_clone/issues/18

zporter replied:
that's an awesome list -- well done!

zporter commented on app/views/posts/_comment.html.erb:
```diff
@@ -0,0 +1,10 @@
+<%= "#{comment.text}" %>
```
`comment.text` is a String, right?

zporter commented on app/views/posts/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+    <%= post.text.truncate(80, {separator: ' '}) %>
+  <% end %>
+  <br />
+  <p>by <%= link_to(post.submitter.name, user_path(post.submitter.name)) %> | posted <%= link_to(time_ago_in_words(post.created_at), post_path(post.id)) %>  ago
```
can `post_path(post.id)` be replaced with `post` in this context? at the very least, the `.id` bit can be dropped.

zporter commented on app/views/posts/new.html.erb:
```diff
@@ -0,0 +1,10 @@
+    <%= render partial: "input", locals: { form: f, field_name: :title } %>
+    <%= render partial: "input", locals: { form: f, field_name: :url } %>
+    <p class="center">OR</p>
+    <%= render partial: "input", locals: { form: f, field_name: :text } %>
```
could the repetition of `render partial: "input", locals: { form: f, field_name: FIELD_NAME }` be cleaned up?

gingin77 replied:
I'm confused by the direction, that you seemed to be hinting at earlier, when you mentioned `<%= input_tag f, field_name %>`. I thought the purpose of the `locals` hash is to pass an instance of the form builder and the attribute to the partial (I was looking at [http://guides.rubyonrails.org/layouts_and_rendering.html#using-partials](url) ). I'm not seeing how a helper could reduce whats in the locals hash - but maybe there is more flexibility allowed than what I have experience with.

zporter replied:
that's a fair point -- all the helper method is really buying you is saved keystrokes and the ability to move that partial in the future without the need to update a whole bunch of references in the code. There's also the extensibility of if you wanted to have a `placeholder` or `hint` text, then you simply make them arguments on the helper method with a sensible default. Using the `locals` makes that trickier to set a default -- you'd end up setting a local var in the template, which isn't the best place to set a Ruby variable.

zporter commented on app/views/posts/show.html.erb:
```diff
@@ -7,14 +7,14 @@
         <tr>
           <td scope="row">link:</td>
-          <td class="col2"><%= link_to post.url %></td>
+          <td class="col2"><%= link_to(post.url, post.url) %></td>
```
I like the explicitness of what you have here, but this could also be written as `link_to nil, post.url`

zporter commented on app/views/posts/show.html.erb:
```diff
@@ -33,22 +33,21 @@
+  <%= f.text_area :text %>
   <br />
-  <%= f.hidden_field :post_id, :value => parent_id %>
+  <%= f.hidden_field :post_id, :value => post.id %>
```
for consistency, convert that hash rocket (`=>`) into the newer Hash syntax (e.g. `value: post`)

zporter commented on app/views/posts/show.html.erb:
```diff
@@ -33,22 +33,21 @@
-  <%= f.submit "add comment" %>
+  <%= f.submit "Direct Comment On Post" %>
+<% end %>
+<% if post.direct_comments.exists? %>
```
good use of `exists?`. I'm curious what the resulting SQL looks like in the console...

zporter commented on config/routes.rb:
```diff
@@ -1,8 +1,12 @@
 
-  resources :submissions
+  resources :posts do
+    resources :comments, only: [:create, :show] do
```
I didn't see a `show` action on the controller nor an associated view. Intentional?

gingin77 replied:
I was thinking about using a comments#show when I was brainstorming ways to handle the error messages, but ended up deciding not to. At some point, a users/:id/comments#index might be desirable but for now I'll delete the :show and wait til I'm ready to use :index before adding it.

gingin77 committed:
Syntax adjustments in a number of places

gingin77 committed:
Remove unused :show from comments resources in routes

gingin77 committed:
Syntax updates in posts_controller and post views - _comments and index

gingin77 committed:
Bring comments back to Gemfile and ensure rspec and data-cleaner gems are still present

gingin77 committed:
Make comments polymorphic, make updates to align with this across app - with exceptions

Specs and factories have been updated - currently passing
Views have been updated to display comments but the form inputs likely need attention, along with looking at what to call permitted params on the create action

gingin77 committed:
WIP: Still working on best way to abstract parent posts, parent comments and new comments in controllers, views, and routes

gingin77 committed:
Remove use of session to hold request.referrer; instead simply redirect to parent ivar without using any route path helpers

Remove '.id' from places where it is not needed because Rails just needs the object

gingin77 committed:
WIP: Add Shoulda-matchers and run some new specs with it

There's a bit of redunancy now between the new should specs and the conventional RSpec style tests I wrote earlier... Need to decide what to keep

gingin77 committed:
Create user model spec to test has_many posts and has_many comments

gingin77 committed:
Update factories/specs to rely on names rather than user_a and user_b

gingin77 committed:
Write passing feature tests for direct and reply comments

Update helper with Devise config info
Modify a view by changing text on button to say 'Submit Reply' instead
of 'Comment on Post'

Set conditional that hides 'Reply to' link on comments if user not auth << this was done while developing the tests

gingin77 committed:
Replace all ivars with let methods in reply comment spec

albaer commented on spec/features/reply_to_comment_spec.rb:
```diff
@@ -0,0 +1,50 @@
+require "rails_helper"
+
+feature "Reply to a Comment" do
```
Cool! I don't think I've ever seen this instead of "describe" before. Where did you get this?

gingin77 replied:
It was in the Capybara docs - "feature is in fact just an alias for describe ..., :type => :feature, background is an alias for before, scenario for it, and given/given! aliases for let/let!, respectively."
https://github.com/teamcapybara/capybara#using-capybara-with-rspec

gingin77 committed:
Remove before wrapper around the let statements in the reply to comment feature test

albaer commented on spec/features/reply_to_comment_spec.rb:
```diff
@@ -0,0 +1,50 @@
+require "rails_helper"
+
+feature "Reply to a Comment" do
+  before :each do
```
I don't think your `let`s need to be in a before block, and it's a little unusual. You could take them out, and if the data needs to be created before the tests run, change them to `let!`

zporter commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+feature "Comment on a Post" do
+  let(:user) { create(:oliver) }
+  let(:post) { create(:url_post,
+    title: "Y Combinator Is Launching A “Grad School” For Booming Startups") }
```
here's how we break `let` statements across multiple lines:

```ruby
let(:post) do
  create :url_post,
         title: "Y Combinator Is Launching A “Grad School” For Booming Startups"
end
```

zporter commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    title: "Y Combinator Is Launching A “Grad School” For Booming Startups") }
+
+  it "allows authenticated user to comment on a post and see new comment on the parent post show view" do
+    submitter = user
```
I'd probably just use `let(:submitter)` or use the `user` variable

albaer replied:
I don't think this is necessary either. Any reason not to just use `user`? Or do `let(:submitter) { create(:oliver) }` on line 4?

zporter commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    submitter = user
+    sign_in submitter
+
+    visit post_path((post).id)
```
I'm curious if `visit post` would work here. if not, `visit post_path(post)` should do it.

zporter commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    sign_in submitter
+
+    visit post_path((post).id)
+    expect(page).to have_http_status(200)
```
can probably remove this line as a 200 can be implied by the content expectations below.

albaer commented on spec/features/reply_to_comment_spec.rb:
```diff
@@ -0,0 +1,48 @@
+
+  it "allows authenticated user to submit a reply comment and get redirected to parent comment show view" do
+    sign_in user
+    current_user = user
```
Is this necessary? Looks like the only reference is on line 30 and that could just call `user`.

zporter commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    expect(page).to have_http_status(200)
+    expect(page).to have_content "Y Combinator Is Launching A “Grad School” For Booming Startups"
+    expect(page).to have_content "contributer:	#{post.submitter.name}"
+    expect(page).to have_selector(:link_or_button, "Direct Comment On Post")
```
unnecessary expectation because of the `click_on` below?

albaer commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+feature "Comment on a Post" do
+  let(:user) { create(:oliver) }
+  let(:post) { create(:url_post,
```
If a block runs onto multiple lines, do/end is preferred over {} brackets.

zporter commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  it "blocks unauthenticated users from submitting comments" do
+    visit post_path(post.id)
```
if `visit post` doesn't work, then `visit post_path(post)` should do it. This will ensure that it continues to work if the `to_param` method on `post` were ever overridden.

gingin77 replied:
Gotcha - `visit post` doesn't work but `visit post_path(post)` does

zporter commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+  it "blocks unauthenticated users from submitting comments" do
+    visit post_path(post.id)
+    expect(page).to have_http_status(200)
```
potentially unnecessary HTTP status check since the content expectations below imply a 200 was received.

albaer reviewed (commented):
Cool! It's awesome that you're adding feature tests. I left a few comments in those files.

zporter commented on spec/features/directly_comment_on_post_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    expect(page).to have_content "You need to sign in or sign up before continuing."
+    expect(page).to have_content "Sign In"
+  end
+end
```
since the steps to fill in a comment are the same and the only difference between these 2 `it` blocks is the authentication at the beginning and the expectations at the end, a method can be defined to house the common steps. Something like:

```ruby
def fill_in_comment
  visit post_path(post)

  expect(page).to have_content "Y Combinator Is Launching A “Grad School” For Booming Startups"
  expect(page).to have_content "contributer:	#{post.submitter.name}"
  expect(page).to have_selector(:link_or_button, "Direct Comment On Post")

  fill_in "comment_text", with: "exciting"
  click_on "Direct Comment On Post"
end
```

gingin77 replied:
Issue #23 created - https://github.com/gingin77/hackernews_clone/issues/23

gingin77 committed:
Adjustments based on PR feedback

Remove unnecessary use of .id from url paths
Remove `expect 200's` cuz they were redundant
Stack and indent long let blocks

gingin77 committed:
Recover original structure of posts/new view; Modify text on button for comment on post

gingin77 committed:
Write test for error handling when blank comments are submitted

Modify comment#create action to pass tests

gingin77 committed:
Refactor an elsif conditional to an else - extra details were not needed

gingin77 committed:
Rename user as current_user; eliminate unneed line

zporter commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,49 @@
+class CommentsController < ApplicationController
+  before_action :authenticate_to_submit, only: :new
+  before_action :authenticate_user!, only: :create
+  before_action :parent, only: :create
```
this is completely unnecessary if `parent` was referenced in the `create` action.

zporter commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,49 @@
+        redirect_to @parent
+      else
+        redirect_to comment_reply_comment_path(@parent)
+      end
```
I honestly feel like this could be just `render :new` here. The text of the parent will be on that view any way so I'm not convinced this conditional is worth sending them back to the post page if the comment came was on a post. thoughts?

zporter commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,49 @@
+    @comment = @parent.comments.build(comment_params)
+    if @comment.save
+      redirect_to @parent
+    elsif !@comment.save
```
this will invoke `.save` again. a simple `else` should work.
